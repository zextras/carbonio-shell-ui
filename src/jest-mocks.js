/*
 * *** BEGIN LICENSE BLOCK *****
 * Copyright (C) 2011-2020 Zextras
 *
 *  The contents of this file are subject to the Zextras EULA;
 * you may not use this file except in compliance with the EULA.
 * You may obtain a copy of the EULA at
 * http://www.zextras.com/zextras-eula.html
 * *** END LICENSE BLOCK *****
 */
import fetch from 'node-fetch';

function soapFetch(api, body) {
	const request = {
		Body: {
			[`${ api }Request`]: body,
		},
	};
	// if (this._csrfToken) {
	// 	request.Header = {
	// 		_jsns: 'urn:zimbra',
	// 		context: {
	// 			csrfToken: this._csrfToken
	// 		}
	// 	};
	// }
	return fetch(
		new URL(`/service/soap/${ api }Request`, 'http://localhost'),
		{
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
			},
			body: JSON.stringify(request),
		},
	)
		.then((r) => r.json())
		.then((resp) => {
			if (resp.Body.Fault) {
				throw new Error(resp.Body.Fault.Reason.Text);
			}
			return resp.Body[`${ api }Response`];
		});
}

export const network = {
	soapFetch,
};

export const fiberChannel = jest.fn();
export const fiberChannelSink = jest.fn();
