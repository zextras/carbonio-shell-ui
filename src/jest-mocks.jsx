/*
 * *** BEGIN LICENSE BLOCK *****
 * Copyright (C) 2011-2020 Zextras
 *
 *  The contents of this file are subject to the Zextras EULA;
 * you may not use this file except in compliance with the EULA.
 * You may obtain a copy of the EULA at
 * http://www.zextras.com/zextras-eula.html
 * *** END LICENSE BLOCK *****
 */
import path from 'path';
import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import fetch from 'node-fetch';
import { MemoryRouter } from 'react-router-dom';
import AppContextWrapper from './mocks/app-context-wrapper';

jest.mock('react-i18next', () => ({
	// this mock makes sure any components using the translate hook can use it without a warning being shown
	useTranslation: () => {
		return {
			t: (str) => str,
			i18n: {
				changeLanguage: () => new Promise(() => {}),
			},
		};
	},
}));

function soapFetch(api, body) {
	const request = {
		Body: {
			[`${api}Request`]: body,
		},
	};
	// if (this._csrfToken) {
	// 	request.Header = {
	// 		_jsns: 'urn:zimbra',
	// 		context: {
	// 			csrfToken: this._csrfToken
	// 		}
	// 	};
	// }
	return fetch(
		new URL(`/service/soap/${api}Request`, 'http://localhost'),
		{
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
			},
			body: JSON.stringify(request),
		},
	)
		.then((r) => r.json())
		.then((resp) => {
			if (resp.Body.Fault) {
				throw new Error(resp.Body.Fault.Reason.Text);
			}
			return resp.Body[`${api}Response`];
		});
}

export const network = {
	soapFetch,
};

export const fiberChannel = jest.fn();
export const fiberChannelSink = jest.fn();

function render(
	ui,
	{
		ctxt,
		reducer,
		preloadedState,
		initialRouterEntries = ['/'],
		...options
	}
) {
	const confPath = path.resolve(
		process.cwd(),
		'zapp.conf.js'
	);
	// eslint-disable-next-line max-len
	// eslint-disable-next-line global-require,import/no-dynamic-require,@typescript-eslint/no-var-requires
	const conf = require(confPath);

	const Wrapper = ({ children }) => (
		<MemoryRouter initialEntries={initialRouterEntries}>
			<AppContextWrapper
				packageName={conf.pkgName}
				packageVersion={conf.version}
				ctxt={ctxt}
				reducer={reducer}
				preloadedState={preloadedState}
			>
				{ children }
			</AppContextWrapper>
		</MemoryRouter>
	);

	return rtlRender(
		ui,
		{
			wrapper: Wrapper,
			...options,
		}
	);
}

export const test = {
	render
};
