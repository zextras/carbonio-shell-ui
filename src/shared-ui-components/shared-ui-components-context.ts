/*
 * *** BEGIN LICENSE BLOCK *****
 * Copyright (C) 2011-2020 Zextras
 *
 *  The contents of this file are subject to the Zextras EULA;
 * you may not use this file except in compliance with the EULA.
 * You may obtain a copy of the EULA at
 * http://www.zextras.com/zextras-eula.html
 * *** END LICENSE BLOCK *****
 */

import { createContext, FC } from 'react';
import { AppPkgDescription } from '../../types';

export type SharedUIComponentsMap = {
	[id: string]: {
		pkg: string;
		versions: AppPkgDescription;
	};
};

export const SharedUIComponentsContext = createContext<SharedUIComponentsMap>({});
