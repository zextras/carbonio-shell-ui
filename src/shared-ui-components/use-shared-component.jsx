import React, { useContext, useMemo } from 'react';
import { get } from 'lodash';
import AppContextProvider from '../app/app-context-provider';
import { SharedUIComponentsContext } from './shared-ui-components-context';

export function useSharedComponent(id, version) {
	const components = useContext(SharedUIComponentsContext);
	const Component = useMemo(() => {
		const C = get(components, [id, 'versions', version]);
		if (C) {
			const SharedComponent = (props) => (
				<AppContextProvider pkg={components[id].pkg}>
					<C {...props} />
				</AppContextProvider>
			);
			return SharedComponent;
		}
		return null;
	}, [components, id, version]);
	return Component;
}
