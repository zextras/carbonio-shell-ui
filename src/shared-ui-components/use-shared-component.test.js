/*
 * *** BEGIN LICENSE BLOCK *****
 * Copyright (C) 2011-2020 Zextras
 *
 *  The contents of this file are subject to the Zextras EULA;
 * you may not use this file except in compliance with the EULA.
 * You may obtain a copy of the EULA at
 * http://www.zextras.com/zextras-eula.html
 * *** END LICENSE BLOCK *****
 */
jest.mock('./shared-ui-components-context-provider');
jest.mock('../app/app-context-provider');
import React, { useContext } from 'react';
import { useSharedComponent } from './use-shared-component';
import SharedUIComponentsContextProvider from './shared-ui-components-context-provider';
import { SharedUIComponentsContext } from './shared-ui-components-context';
import { testUtils } from '../jest-mocks';
import AppContextProvider from '../app/app-context-provider';

describe('useSharedComponent hook', () => {
	test('get a shared component', () => {
		const ComponentClass11 = jest.fn().mockImplementation(() => 'Scope-1-Component-Package-1');
		const ComponentClass21 = jest.fn().mockImplementation(() => 'Scope-2-Component-Package-1');
		const ComponentClass22 = jest.fn().mockImplementation(() => 'Scope-2-Component-Package-2');
		const ComponentClass32 = jest.fn().mockImplementation(() => 'Scope-3-Component-Package-2');
		SharedUIComponentsContextProvider.mockImplementationOnce(({ children }) => (
			<SharedUIComponentsContext.Provider
				value={{
					'component-4': {
						pkg: 'com_example_package_2',
						versions: {
							'1': ComponentClass22,
							'2': ComponentClass22
						}
					},
					'component-3': {
						pkg: 'com_example_package_2',
						versions: {
							'1': ComponentClass32,
							'2': ComponentClass32
						}
					},
					'component-1': {
						pkg: 'com_example_package_1',
						versions: {
							'1': ComponentClass11,
							'2': ComponentClass11
						}
					},
					'component-2': {
						pkg: 'com_example_package_1',
						versions: {
							'1': ComponentClass21,
							'2': ComponentClass21
						}
					}
				}}
			>
				{children}
			</SharedUIComponentsContext.Provider>
		));
		AppContextProvider.mockImplementationOnce(({ children, ...props }) => (
			<>
				AppContextProvider props: {JSON.stringify(props)}
				{children}
			</>
		));
		const tester = jest.fn();

		function Tester() {
			const ctxt = useContext(SharedUIComponentsContext);
			const C = useSharedComponent('component-4', '1');
			console.log(C);
			tester(ctxt);
			return <C/>;
		}

		const { container } = testUtils.render(
			<SharedUIComponentsContextProvider>
				<Tester />
			</SharedUIComponentsContextProvider>
		);

		expect(container).toMatchSnapshot();
		expect(tester).toHaveBeenCalledTimes(1);
	});
});
