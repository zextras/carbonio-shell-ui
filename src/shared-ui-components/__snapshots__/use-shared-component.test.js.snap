// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSharedComponent hook get a shared component 1`] = `
<div>
  AppContextProvider props: 
  {"pkg":"com_example_package_2"}
  Scope-2-Component-Package-2
</div>
`;
